packages/core/src/services/chains/ChainStreamConsumer/index.ts

```diff
@@ -0,0 +1,173 @@
+import { MessageRole } from '@latitude-data/compiler'
+
+import {
+  ChainEvent,
+  ChainEventTypes,
+  ChainStepResponse,
+  RunErrorCodes,
+  StreamEventTypes,
+  StreamType,
+} from '../../../constants'
+import { objectToString } from '../../../helpers'
+import { Config } from '../../ai'
+import { ChainError } from '../ChainErrors'
+import { ValidatedStep } from '../ChainValidator'
+
+export function parseResponseText(response: ChainStepResponse<StreamType>) {
+  if (response.streamType === 'object') return response.text || ''
+
+  const text = response.text
+  const hasTextResponse = text && text.length > 0
+  if (hasTextResponse) return text
+
+  if (!response.toolCalls.length) return text ?? ''
+
+  const toolsString = objectToString(
+    response.toolCalls,
+    'Error parsing tool calls',
+  )
+  return `Tool calls requested: ${toolsString}`
```

Comment:
@geclos ☝️ I ported your changes here, let me know if you see it ok

Reply:
look coolio